v2/internal/reflecthelpers/reflect_helpers.go

```diff
@@ -282,3 +282,74 @@ func getItemsField(listPtr client.ObjectList) (reflect.Value, error) {
 
 	return itemsField, nil
 }
+
+// SetProperty sets the property on the provided object to the provided value.
+// obj is the object to modify.
+// propertyPath is a dot-separated path to the property to set.
+// value is the value to set the property to.
+// Returns an error if any of the properties in the path do not exist, if the property is not settable,
+// or if the value provided is incompatible.
+func SetProperty(obj any, propertyPath string, value any) error {
+	steps := strings.Split(propertyPath, ".")
+	return setPropertyCore(obj, steps, value)
+}
+
+func setPropertyCore(obj any, propertyPath []string, value any) (err error) {
+	// Get the underlying object we need to modify
+	subject := reflect.ValueOf(obj)
+
+	// Dereference pointers
+	if subject.Kind() == reflect.Ptr {
+		subject = subject.Elem()
+	}
+
+	// Check we have a struct
+	if subject.Kind() != reflect.Struct {
+		return errors.Errorf("provided object was not a struct, was %s", subject.Kind())
+	}
+
+	// Get the field we need to modify
+	field := subject.FieldByName(propertyPath[0])
+
+	// Check the field exists
+	if field == (reflect.Value{}) {
+		return errors.Errorf("provided object did not have a field named %s", propertyPath[0])
+	}
+
+	// If this is not the last property in the path, we need to recurse
+	if len(propertyPath) > 1 {
+		if field.Kind() == reflect.Ptr {
+			// Field is a pointer; initialize it if needed, then pass the pointer recursively
+			if field.IsNil() {
+				newValue := reflect.New(field.Type().Elem())
+				field.Set(newValue)
+			}
+
+			return errors.Wrapf(
+				setPropertyCore(field.Interface(), propertyPath[1:], value),
+				"failed to set property %s",
+				propertyPath[0])
+		}
+
+		// Field is not a pointer, so we need to pass the address of the field recursively
+		return errors.Wrapf(
+			setPropertyCore(field.Addr().Interface(), propertyPath[1:], value),
+			"failed to set property %s",
+			propertyPath[0])
+	}
+
+	// If this is the last property in the path, we need to set the value, if we can
+	if !field.CanSet() {
+		return errors.Errorf("field %s was not settable", propertyPath[0])
+	}
+
+	// Catch any panic and turn it into an error return
+	defer func() {
```

Comment:
Don't you need to put this at the front of the function?
https://go.dev/play/p/nF13opHFOfO

Otherwise I don't think it'll be reached if anything above it panics?

Reply:
As discussed, I've moved the recover to the top of the method.